upgrade: use high priority txn's to update the cluster version

fqazi opened the pull request:
Previously, it was possible for the leasing subsystem to starve out attempts to set the cluster version during upgrades, since the leasing subsystem uses high priority txn for renewals. To address this, this patch makes the logic to set the cluster version high priority so it can't be pushed out by lease renewals.

Fixes: #113908

Release note (bug fix): Addressed a bug that could cause cluster version finalization to get starved out by descriptor lease renewals on larger clusters.

fqazi committed:
upgrade: use high priority txn's to update the cluster version

Previously, it was possible for the leasing subsystem to starve out
attempts to set the cluster version during upgrades, since the leasing
subsystem uses high priority txn for renewals. To address this, this
patch makes the logic to set the cluster version high priority so
it can't be pushed out by lease renewals.

Fixes: #113908

Release note (bug fix): Addressed a bug that could cause cluster version
finalization to get starved out by descriptor lease renewals on larger
clusters.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/113996)
<!-- Reviewable:end -->

nvb reviewed (commented):
Thanks @fqazi!

In addition to the test that you're adding here, there are two other thing that would give us more confidence in the efficacy and safety of this change:
1. we demonstrate that it fixes the reproduction presented in https://github.com/cockroachlabs/support/issues/2685#issuecomment-1788125904.
2. we determine all of the keys that the upgrade txn that we're marking as high priority could read from and write to. The code itself makes it look like we're only reading from the version key and then writing to it. But are there any other reads or writes that are performed on behalf of the txn (now at high priority) due to the use of an internal executor?

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-:-NikQM056o_u1j7DweMx:b-cvs69v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @fqazi)

___
*[`pkg/sql/set_cluster_setting.go` line 621 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NikPWBW7G_wYCrCMFWh:-NikPWBW7G_wYCrCMFWi:by52mr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/829243e44382cd7d935cb80a8c767143ff95e74a/pkg/sql/set_cluster_setting.go#L621)):*
> ```Go
> 		return db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
> 			// Run the version bump inside the upgrade as high priority, since
> 			// lease manager ends up locking the version row inside the settings
> ```

s/locking/reading/

"with high priority"
___
*[`pkg/sql/set_cluster_setting.go` line 624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NikR5hCEL2H-JEgtjBH:-NikR5hCEL2H-JEgtjBI:bn9xc1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/829243e44382cd7d935cb80a8c767143ff95e74a/pkg/sql/set_cluster_setting.go#L624)):*
> ```Go
> 			// table when refreshing leases. On larger clusters or with a large number
> 			// of descriptors its possible for normal transactions to be starved
> 			// (#95227)
> ```

nit: punctuation.
___
*[`pkg/sql/set_cluster_setting.go` line 624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NikRB8L3YGawLrTrWWx:-NikRB8L3YGawLrTrWWy:by7bsob) ([raw file](https://github.com/cockroachdb/cockroach/blob/829243e44382cd7d935cb80a8c767143ff95e74a/pkg/sql/set_cluster_setting.go#L624)):*
> ```Go
> 			// table when refreshing leases. On larger clusters or with a large number
> 			// of descriptors its possible for normal transactions to be starved
> 			// (#95227)
> ```

Let's also add a paragraph about why this is safe (see my comment above).

Assuming that we're only reading from the version key and then writing to it, changing all of these upgrade transactions from normal priority to high priority should not alter how they interact with each other — two transactions at the same priority interact the same with each other, regardless of what that priority is.

So the only effects this will have are:
1. the transaction will now abort any other transactions that write to the version key with normal priority (there aren't any, right?)
2. the transaction will now block any other transaction that reads from the version key. That's kind of the point.
3. Depending on whether the internal executor is hiding any other reads or writes, this may interact differently with other txns for those.
___
*[`pkg/sql/set_cluster_setting.go` line 660 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NikS99X1t3hp1_hDS7T:-NikS99X1t3hp1_hDS7U:b-52yu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/829243e44382cd7d935cb80a8c767143ff95e74a/pkg/sql/set_cluster_setting.go#L660)):*
> ```Go
> 				ctx, "update-setting", txn.KV(),
> 				sessiondata.RootUserSessionDataOverride,
> 				`UPSERT INTO system.settings (name, value, "lastUpdated", "valueType") VALUES ($1, $2, now(), $3)`,
> ```

Could this use of SQL ever result in a lease acquisition? Does that risk a deadlock, if the lease acquisition gets stuck on the version upgrade's write?
___
*[`pkg/upgrade/upgrades/version_starvation_test.go` line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NikPlAO1Y_H_hC5BJH6:-NikPlAP1bRnVEZ3GEzT:b-2irp6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/829243e44382cd7d935cb80a8c767143ff95e74a/pkg/upgrade/upgrades/version_starvation_test.go#L32)):*
> ```Go
> // the cluster version is done with a high priority txn and cannot
> // be pushed out. Previously, this would be normal priority and
> // get pushed by the leasing code. Note: This test just confirms
> ```

"by the leasing code, leading to starvation when leases were acquired with sufficiently high frequency"


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @nvanbenschoten)

___
*[`pkg/sql/set_cluster_setting.go` line 660 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NikS99X1t3hp1_hDS7T:-NikrT5g44tqm11-9fnq:b1lfpwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/829243e44382cd7d935cb80a8c767143ff95e74a/pkg/sql/set_cluster_setting.go#L660)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could this use of SQL ever result in a lease acquisition? Does that risk a deadlock, if the lease acquisition gets stuck on the version upgrade's write?
</blockquote></details>

Oh good point that is a potential danger here, if a lease renewal is needed on the system.settings table. Specifically a synchronous renewal was required then we can deadlock.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @nvanbenschoten)

___
*[`pkg/sql/set_cluster_setting.go` line 624 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NikRB8L3YGawLrTrWWx:-NioP6Ts19Gnn574Hw4I:b-kj8lyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/829243e44382cd7d935cb80a8c767143ff95e74a/pkg/sql/set_cluster_setting.go#L624)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's also add a paragraph about why this is safe (see my comment above).

Assuming that we're only reading from the version key and then writing to it, changing all of these upgrade transactions from normal priority to high priority should not alter how they interact with each other — two transactions at the same priority interact the same with each other, regardless of what that priority is.

So the only effects this will have are:
1. the transaction will now abort any other transactions that write to the version key with normal priority (there aren't any, right?)
2. the transaction will now block any other transaction that reads from the version key. That's kind of the point.
3. Depending on whether the internal executor is hiding any other reads or writes, this may interact differently with other txns for those.
</blockquote></details>

Clarified why this should be safe:
// Run the version bump inside the upgrade as high priority, since  
// lease manager ends up reading the version row (with high priority)  
// inside the settings table when refreshing leases. On larger clusters  
// or with a large number of descriptors its possible for normal  
// transactions to be starved  (#95227).  
// This is safe because we expected this transaction only do the following:  
// 1) We expect this transaction to only read and write to the  
//    version key in the system.settings table.  
// 2) Reads from the system.sql\_instances table to confirm all SQL servers  
//    have been upgraded in multi-tenant environments.  
// 3) Other transactions will use a normal priority and get pushed out by  
//    this one.  
// 4) There is a potential danger to conflict with lease renewal of the  
//    settings table and the upgrade, but this cannot occur in practice,  
//    because the lease will always be acquired before we try to write to  
//    the settings table.

If we want to be paranoid we could disable using leased descriptors in the planner.
___
*[`pkg/sql/set_cluster_setting.go` line 660 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NikS99X1t3hp1_hDS7T:-NioP7Bc6jZWVZvMmIP6:b-xc3eky) ([raw file](https://github.com/cockroachdb/cockroach/blob/829243e44382cd7d935cb80a8c767143ff95e74a/pkg/sql/set_cluster_setting.go#L660)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Oh good point that is a potential danger here, if a lease renewal is needed on the system.settings table. Specifically a synchronous renewal was required then we can deadlock.
</blockquote></details>

Turns out this is safe, with some experimentation I was able to confirm this is a non-issue, since we will always get the lease to the setting table before writing.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
2) I looked into (2) and the only other operation is a read to the sqlinstances table.

I'm going to run (1) locally to reproduce the scenario and confirm the fix works

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-:-NioPW4GDHFMo9GMbr0f:b-juwo2h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 6 of 6 files at r3, 3 of 3 files at r4, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @fqazi)

___
*[`-- commits` line 29 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8GzJuCZp7-fiaqpLG):*
"simulate"?
___
*[`pkg/sql/set_cluster_setting.go` line 623 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8FWI50egodniOhBZg:-Nj8FWI50egodniOhBZh:bxwzyn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b5712a7454d772369fe3800a038135b67ffb14/pkg/sql/set_cluster_setting.go#L623)):*
> ```Go
> 			// lease manager ends up reading the version row (with high priority)
> 			// inside the settings table when refreshing leases. On larger clusters
> 			// or with a large number of descriptors its possible for normal
> ```

nit: "it's"
___
*[`pkg/sql/set_cluster_setting.go` line 626 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8FvJNB-GbYf7UcUda:-Nj8FvJNB-GbYf7UcUdb:b-ksz7gt) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b5712a7454d772369fe3800a038135b67ffb14/pkg/sql/set_cluster_setting.go#L626)):*
> ```Go
> 			// transactions to be starved  (#95227).
> 			// This is safe because we expected this transaction only do the following:
> 			// 1) We expect this transaction to only read and write to the
> ```

How does the use of `WithSkipDescriptorCache` change this behavior? Are there new keys that we read from and write to?
___
*[`pkg/sql/set_cluster_setting.go` line 630 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8Gm013rm-7bc2xJ7M:-Nj8Gm013rm-7bc2xJ7N:b-x571wf) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b5712a7454d772369fe3800a038135b67ffb14/pkg/sql/set_cluster_setting.go#L630)):*
> ```Go
> 			// 2) Reads from the system.sql_instances table to confirm all SQL servers
> 			//    have been upgraded in multi-tenant environments.
> 			// 3) Other transactions will use a normal priority and get pushed out by
> ```

Which other transactions is this referring to? Other transactions that write to the version key? Do such transactions exist?
___
*[`pkg/sql/set_cluster_setting.go` line 634 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8GIfXAD6tHp65UbjR:-Nj8GIfXAD6tHp65UbjS:bjsk7jk) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b5712a7454d772369fe3800a038135b67ffb14/pkg/sql/set_cluster_setting.go#L634)):*
> ```Go
> 			// 4) There is a potential danger to conflict with lease renewal of the
> 			//    settings table and the upgrade, but to avoid this, we intentionally
> 			//		force the internal executor to avoid leased descriptors.
> ```

nit: indentation looks off here.
___
*[`pkg/sql/set_cluster_setting.go` line 686 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8FtVD3cxCOsLgMBrH:-Nj8FtVD3cxCOsLgMBrI:b-rqf3zn) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b5712a7454d772369fe3800a038135b67ffb14/pkg/sql/set_cluster_setting.go#L686)):*
> ```Go
> 			return err
> 		},
> 			isql.WithSkipDescriptorCache())
> ```

Could we give this a comment? Specifically, why does this help? If we never perform leasing then this transaction will never get stuck waiting on a leasing transaction. Does it have to perform other reads and writes to compensate? If so, how do those operations interact with other transactions?
___
*[`pkg/sql/catalog/lease/kv_writer.go` line 71 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8HCVCD6j_Twpx0uJ0:-Nj8HCVCD6j_Twpx0uJ1:b6fee5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ac22a2a7bfd15fa210783f446219020ed69ad8d/pkg/sql/catalog/lease/kv_writer.go#L71)):*
> ```Go
> ) (settingswatcher.VersionGuard, error) {
> 	targetVersion := clusterversion.TODO_Delete_V23_1_SystemRbrCleanup
> 	if w.knobs.ForceVersionGuardReads {
> ```

Would it be cleaner to push this knob one level deeper and consult it directly in `SettingsWatcher.MakeVersionGuard`? So we would have logic in that function that looks like: `if activeVersion.IsActive(maxGate) || forceVersionGuardReads {`


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @nvanbenschoten)

___
*[`pkg/sql/set_cluster_setting.go` line 626 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8FvJNB-GbYf7UcUda:-Nj8Inp5-3eDag7ZKfR3:b6981go) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b5712a7454d772369fe3800a038135b67ffb14/pkg/sql/set_cluster_setting.go#L626)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

How does the use of `WithSkipDescriptorCache` change this behavior? Are there new keys that we read from and write to?
</blockquote></details>

// 1) We expect this transaction to only read and write to the  
//    version key in the system.settings table. It will also do high  
//    priority reads from the system.namespace and system.descriptor tables  
//    because the WithSkipDescriptorCache below (this will push out schema  
//    changes during the upgrade).
___
*[`pkg/sql/set_cluster_setting.go` line 630 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8Gm013rm-7bc2xJ7M:-Nj8IzN0A3FGngrrlFff:b-rqy0bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b5712a7454d772369fe3800a038135b67ffb14/pkg/sql/set_cluster_setting.go#L630)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Which other transactions is this referring to? Other transactions that write to the version key? Do such transactions exist?
</blockquote></details>

I was thinking of other transactions involving the system database getting pushed if they involve the system database descriptor:

// 3) Other transactions will use a normal priority and get pushed out by  
//    this one, if they involve schema changes on the system database  
//    descriptor (highly unlikely).
___
*[`pkg/sql/set_cluster_setting.go` line 686 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8FtVD3cxCOsLgMBrH:-Nj8JCFD-sPy12YWmL5F:b2n1a9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b5712a7454d772369fe3800a038135b67ffb14/pkg/sql/set_cluster_setting.go#L686)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could we give this a comment? Specifically, why does this help? If we never perform leasing then this transaction will never get stuck waiting on a leasing transaction. Does it have to perform other reads and writes to compensate? If so, how do those operations interact with other transactions?
</blockquote></details>

I clarified this above, but I'll add another comment stating why this is needed here:

// We will intentionally skip the leasing layer to update the  
// system.settings table to avoid a deadlock with the leasing layer.  
// The leasing layer will read the version key in system.settings with  
// a high priority before any renewal. The update code above will need a  
// lease on system.settings and potentially system.sqlinstances. If any of  
// leases required expire, we could deadlock with the renewal being  
// blocked on the version key. To get around this, we will force the  
// transaction to directly make round trips to read descriptors instead  
// of using the leasing cache.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 6 of 6 files at r6, 9 of 9 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @fqazi)

___
*[`pkg/sql/set_cluster_setting.go` line 686 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8FtVD3cxCOsLgMBrH:-Nj8wQP43bIZe0MJaku_:baz6k0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b5712a7454d772369fe3800a038135b67ffb14/pkg/sql/set_cluster_setting.go#L686)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I clarified this above, but I'll add another comment stating why this is needed here:

// We will intentionally skip the leasing layer to update the  
// system.settings table to avoid a deadlock with the leasing layer.  
// The leasing layer will read the version key in system.settings with  
// a high priority before any renewal. The update code above will need a  
// lease on system.settings and potentially system.sqlinstances. If any of  
// leases required expire, we could deadlock with the renewal being  
// blocked on the version key. To get around this, we will force the  
// transaction to directly make round trips to read descriptors instead  
// of using the leasing cache.
</blockquote></details>

Thanks for spelling this out. The high-priority read on the system.sqlinstances and system.descriptors tables makes me vaguely concerned that this may run into trouble with https://github.com/cockroachdb/cockroach/issues/104728. We'll want to make sure that this change is minimizing risk. Expanding the footprint of this transaction while making it high priority is working against this goal.

This makes me wonder how difficult it would be to have the upgrade transaction read and write the settings key directly, without going through SQL. The leasing transaction is already doing this with `GetClusterVersionFromStorage`. Would it be difficult to do the same here?
___
*[`pkg/server/settingswatcher/version_guard_test.go` line 145 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8vSvJ2oOt6xvtynHP:-Nj8vSvJ2oOt6xvtynHQ:b-h9thht) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c5aa88b55bc2b5e2bd80dd7ce5126338af5c04a/pkg/server/settingswatcher/version_guard_test.go#L145)):*
> ```Go
> 			watcher := settingswatcher.New(nil, s.Codec(), settings, nil, nil, nil)
> 			require.NoError(t, kvDB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
> 				guard, err := watcher.MakeVersionGuard(ctx, txn, maxVersion, false /*force versin guard reads*/)
> ```

nit: the comment here is off.


<!-- Sent from Reviewable.io -->

chrisseto commented on pkg/sql/isql/options.go:
```diff
@@ -50,6 +50,12 @@ func WithSessionData(sd *sessiondata.SessionData) Option {
 	return (*sessionDataOption)(sd)
 }
 
+// WithSkipDescriptorCache forces the internal executor to avoid all leasing
```
Could you add a short not about when or why this option might be used and any warnings that users should consider if they do opt to use this option?

chrisseto commented on pkg/sql/set_cluster_setting.go:
```diff
@@ -617,6 +617,28 @@ func setVersionSetting(
 			return err
 		}
 		return db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			// Run the version bump inside the upgrade as high priority, since
+			// lease manager ends up reading the version row (with high priority)
+			// inside the settings table when refreshing leases. On larger clusters
+			// or with a large number of descriptors it's possible for normal
+			// transactions to be starved  (#95227).
```
Could we add some reference numbers to "large" here? Doesn't have to be exact, just within the ballpark as large is quite ambiguous.

The wording here feels a bit difficult to digest, in particular, "normal transactions" made me think about non-system transactions but I think the intended meaning is non-high priority transactions.

What would you think about (With appropriate line wrapping and appropriate reference numbers):
```suggestion
			// On clusters with a large number of descriptors (> 10k) or nodes (>1k), non-priority transactions reading the version row can be staved. This is due to the lease manager reading the version row at high priority (#95227).
			// Run the version bump inside the upgrade as a high priority txn to avoid being bullied out by the lease manager.
```

chrisseto commented on pkg/upgrade/upgrades/version_starvation_test.go:
```diff
@@ -0,0 +1,124 @@
+// TestLeasingClusterVersionStarvation validates that setting
+// the cluster version is done with a high priority txn and cannot
+// be pushed out. Previously, this would be normal priority and
+// get pushed by the leasing by the leasing code, leading to starvation
```
```suggestion
// get pushed by the leasing, leading to starvation
```

chrisseto commented on pkg/upgrade/upgrades/version_starvation_test.go:
I'm not sure which is the correct question to ask here: Why is this in a retry loop or why is there a `Commit` call after an infinite loop?

Are we expecting the select for update query to receive the txn retry error?

chrisseto commented on pkg/sql/set_cluster_setting.go:
```diff
@@ -617,6 +617,28 @@ func setVersionSetting(
+			// inside the settings table when refreshing leases. On larger clusters
+			// or with a large number of descriptors it's possible for normal
+			// transactions to be starved  (#95227).
+			// This is safe because we expected this transaction only do the following:
```
I think this is a me issue but I don't see how these items say using a high priority txn is safe? I also don't understand how it may be unsafe 😅

chrisseto reviewed (commented):
Mostly just questions and some comment nits.

The change makes sense to me but I don't think I can vouch for it being side effect free so I'll leave an approval to Nathan.

Will the session based leasing work fix this issue more reliably? If so, can we add a note or link to either a commit message or comment?

fqazi reviewed (commented):
They would eliminate one of the causes, unfortunately that migration will run into this, so it will temporarily invoke the same wrath :eyes:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-:-Nj9lOsGC0qHVWBFmd6O:b-9uyx52)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto, @dt, and @nvanbenschoten)

___
*[`pkg/sql/set_cluster_setting.go` line 686 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-Nj8FtVD3cxCOsLgMBrH:-Nj8zhA-B84mmDyHCOFf:b-93ysm) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b5712a7454d772369fe3800a038135b67ffb14/pkg/sql/set_cluster_setting.go#L686)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Thanks for spelling this out. The high-priority read on the system.sqlinstances and system.descriptors tables makes me vaguely concerned that this may run into trouble with https://github.com/cockroachdb/cockroach/issues/104728. We'll want to make sure that this change is minimizing risk. Expanding the footprint of this transaction while making it high priority is working against this goal.

This makes me wonder how difficult it would be to have the upgrade transaction read and write the settings key directly, without going through SQL. The leasing transaction is already doing this with `GetClusterVersionFromStorage`. Would it be difficult to do the same here?
</blockquote></details>

We can directly do the version bump using KV calls. The system.sqlinstances comes from the upgrade manager hook and deeply nested so we can't easily remove that (that would only impact serverless).
___
*[`pkg/sql/set_cluster_setting.go` line 624 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/113996#gh-1391680597:-Nj9lNaA2OEkflwn5hIb:b-aquz6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c5aa88b55bc2b5e2bd80dd7ce5126338af5c04a/pkg/sql/set_cluster_setting.go#L624)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Could we add some reference numbers to "large" here? Doesn't have to be exact, just within the ballpark as large is quite ambiguous.

The wording here feels a bit difficult to digest, in particular, "normal transactions" made me think about non-system transactions but I think the intended meaning is non-high priority transactions.

What would you think about (With appropriate line wrapping and appropriate reference numbers):
```suggestion
			// On clusters with a large number of descriptors (> 10k) or nodes (>1k), non-priority transactions reading the version row can be staved. This is due to the lease manager reading the version row at high priority (#95227).
			// Run the version bump inside the upgrade as a high priority txn to avoid being bullied out by the lease manager.
```
</blockquote></details>

// On complex clusters with a large number of descriptors (> 500) and  
// multi-region nodes (> 9), normal priority transactions reading/updating // the version row can be starved. This is due to the lease manager reading // the version row at high priority, when refreshing leases (#95227), with // a complex cluster this traffic will continuous.
___
*[`pkg/sql/set_cluster_setting.go` line 625 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/113996#gh-1391694207:-Nj9gBLxEBsn3k2WCxGB:b-di6yeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c5aa88b55bc2b5e2bd80dd7ce5126338af5c04a/pkg/sql/set_cluster_setting.go#L625)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

I think this is a me issue but I don't see how these items say using a high priority txn is safe? I also don't understand how it may be unsafe 😅 
</blockquote></details>

Clarified this with:

// This is safe from deadlocks / starvation because we expected this  
// transaction only do the following:
___
*[`pkg/sql/isql/options.go` line 53 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/113996#gh-1391659441:-Nj9cLHREfjplEr8t0uM:b-ubf72n) ([raw file](https://github.com/cockroachdb/cockroach/blob/86bbda1a4ce0b02d296c1ea7d9510913da6d85c7/pkg/sql/isql/options.go#L53)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Could you add a short not about when or why this option might be used and any warnings that users should consider if they do opt to use this option?
</blockquote></details>

Updated to:

WithSkipDescriptorCache forces the internal executor to avoid all leasing infrastructure, which will force descriptors to be read from KV. This can  incur network latencies when accessing descriptors.
___
*[`pkg/upgrade/upgrades/version_starvation_test.go` line 34 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/113996#gh-1391685494:-Nj9cfjc26dd1ojVm5Ui:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c5aa88b55bc2b5e2bd80dd7ce5126338af5c04a/pkg/upgrade/upgrades/version_starvation_test.go#L34)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

```suggestion
// get pushed by the leasing, leading to starvation
```
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/version_starvation_test.go` line 100 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/113996#gh-1391692134:-Nj9cl2e7OkedQ_zL0iY:b-ik7e8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c5aa88b55bc2b5e2bd80dd7ce5126338af5c04a/pkg/upgrade/upgrades/version_starvation_test.go#L100)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

I'm not sure which is the correct question to ask here: Why is this in a retry loop or why is there a `Commit` call after an infinite loop?

Are we expecting the select for update query to receive the txn retry error?
</blockquote></details>

Ooh good catch, yes the commit isn't needed. We expect the select to get pushed and hit an error eventually.


<!-- Sent from Reviewable.io -->

daniel-crlabs commented:
Do we know when this issue will be fixed? We have a high visibility customer who is severely impacted by this. No one seems to be assigned to this, so I'm cc'ing everyone who has commented on it:

@fqazi @nvanbenschoten @andreimatei @rupertharwood-crl

chrisseto reviewed (commented):
> unfortunately that migration will run into this

Oof, that's rough.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though I'll leave the final sign off to @nvanbenschoten 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-:-NjNcnbs6oBado_edObm:ba54i5n)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 15 of 15 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @fqazi)

___
*[`pkg/sql/set_cluster_setting.go` line 625 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjNpIy99BCYXZjqMKqR:-NjNpIy99BCYXZjqMKqS:b-3zzt4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7455a4bd3664424b49845dae6b1f00c63ebebfc4/pkg/sql/set_cluster_setting.go#L625)):*
> ```Go
> 			return err
> 		}
> 		return db.KV().Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
> ```

Could you add a comment about why this goes out of its way to use raw KV?
___
*[`pkg/sql/set_cluster_setting.go` line 633 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjNoSGo7AtCWv6MiXMp:-NjNoSGo7AtCWv6MiXMq:b-bypwle) ([raw file](https://github.com/cockroachdb/cockroach/blob/7455a4bd3664424b49845dae6b1f00c63ebebfc4/pkg/sql/set_cluster_setting.go#L633)):*
> ```Go
> 			// Run the version bump inside the upgrade as a high priority txn to avoid
> 			// being starved out by the lease manager.
> 			// Run the version bump inside the upgrade as high priority, since
> ```

nit: some redundancy here. "Run the version bump inside the upgrade as" twice.
___
*[`pkg/sql/set_cluster_setting.go` line 682 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjNqqq57-LnI7T1p18-:-NjNqqq57-LnI7T1p180:b-9u6sbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7455a4bd3664424b49845dae6b1f00c63ebebfc4/pkg/sql/set_cluster_setting.go#L682)):*
> ```Go
> 			}
> 			// Encode the setting value to write out the updated version.
> 			var tuple []byte
> ```

This all looks correct to me, but the logic feels easy to get wrong. It also feels like it could drift from SQL over time. Do we feel like we have sufficient testing around this upgrade code to give us confidence that we're writing SQL-decodable bytes?
___
*[`pkg/sql/set_cluster_setting.go` line 682 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjNrMvy7JB3f6XNNHQn:-NjNrMvy7JB3f6XNNHQo:bbflky6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7455a4bd3664424b49845dae6b1f00c63ebebfc4/pkg/sql/set_cluster_setting.go#L682)):*
> ```Go
> 			}
> 			// Encode the setting value to write out the updated version.
> 			var tuple []byte
> ```

On ^ note, consider pulling this into a separate function and unit testing it.
___
*[`pkg/sql/set_cluster_setting.go` line 707 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjNppYj1tojubDpvL-K:-NjNppYj1tojubDpvL-L:b-m09b5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7455a4bd3664424b49845dae6b1f00c63ebebfc4/pkg/sql/set_cluster_setting.go#L707)):*
> ```Go
> 			newValue.SetTuple(tuple)
> 			if row.Value != nil {
> 				if err := txn.CPut(ctx, row.Key, newValue, row.Value.TagAndDataBytes()); err != nil {
> ```

This is all in a serializable transaction, so is there a need for the CPut? I might be missing something.


<!-- Sent from Reviewable.io -->

fqazi committed:
settingswatcher: add code for encoding system.setting rows

To be able to have minimal high priority txns for bumping cluster
version, we need the ability to write the setting table using KV calls.
This patch adds, the functions EncodeSettingKey and EncodeSettingValue
to allow directly generate values for Put/Get calls.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt and @nvanbenschoten)

___
*[`pkg/sql/set_cluster_setting.go` line 682 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjNqqq57-LnI7T1p18-:-NjPN51N--jCZhGfNA8I:bwtxcdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7455a4bd3664424b49845dae6b1f00c63ebebfc4/pkg/sql/set_cluster_setting.go#L682)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This all looks correct to me, but the logic feels easy to get wrong. It also feels like it could drift from SQL over time. Do we feel like we have sufficient testing around this upgrade code to give us confidence that we're writing SQL-decodable bytes?
</blockquote></details>

Done

Added unit tests as well.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thank you for the hard work on pushing this along @fqazi. I think we can go ahead an merge this once the remaining comments are addressed.

I do wonder if the backport would benefit from another look from someone else who has worked in this code recently, just to provide additional confidence that we aren't missing anything. @ajstorm or @JeffSwenson would either of you two (or both!) like to provide a review?

> Reviewed 8 of 8 files at r12, 4 of 4 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-:-NjQSmrv2alZdO8wRxGz:bodc8y0)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dt and @fqazi)

___
*[`-- commits` line 4 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjQQqdW0GrrXwR6CQkb):*
"minimimal"
___
*[`-- commits` line 5 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjQQsaG-v04SZ9DXcDO):*
"s/writoe/write to/"
___
*[`pkg/server/settingswatcher/setting_encoder.go` line 27 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjQRtu39zT3XZdl7gpa:-NjQRtu39zT3XZdl7gpb:b-h605f4) ([raw file](https://github.com/cockroachdb/cockroach/blob/64d770aaae93d666fc480de62cf35c1af2e31843/pkg/server/settingswatcher/setting_encoder.go#L27)):*
> ```Go
> // EncodeSettingKey encodes a key for the system.settings table, which
> // can be used for direct KV operations.
> func EncodeSettingKey(codec keys.SQLCodec, setting string) []byte {
> ```

Should we call this from `GetClusterVersionFromStorage`?
___
*[`pkg/server/settingswatcher/setting_encoder.go` line 29 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjQS1NR2cB8dHEvZqN2:-NjQS1NR2cB8dHEvZqN3:b-alxims) ([raw file](https://github.com/cockroachdb/cockroach/blob/64d770aaae93d666fc480de62cf35c1af2e31843/pkg/server/settingswatcher/setting_encoder.go#L29)):*
> ```Go
> func EncodeSettingKey(codec keys.SQLCodec, setting string) []byte {
> 	indexPrefix := codec.IndexPrefix(keys.SettingsTableID, uint32(1))
> 	return encoding.EncodeUvarintAscending(encoding.EncodeStringAscending(indexPrefix, "version"), uint64(0))
> ```

Was this intending to pass `setting`, instead of `"version"`?
___
*[`pkg/sql/set_cluster_setting.go` line 682 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjNqqq57-LnI7T1p18-:-NjQSjNR5Z_h0-V9IOLx:b2nm9wr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7455a4bd3664424b49845dae6b1f00c63ebebfc4/pkg/sql/set_cluster_setting.go#L682)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done

Added unit tests as well.
</blockquote></details>

This turned out nicely.
___
*[`pkg/sql/set_cluster_setting.go` line 636 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjQSd8F2bsAutu1zc0k:-NjQSd8F2bsAutu1zc0l:bunr24o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf31fbd848ccd58ea3fadd8be33af3fccef6393/pkg/sql/set_cluster_setting.go#L636)):*
> ```Go
> 			// 1) We expect this transaction to only read and write to the
> 			//    version key in the system.settings table. To achieve the smallest
> 			//    possible txn and extra to other tables we are going to
> ```

"and extra to"

Something sounds off here.
___
*[`pkg/server/settingswatcher/setting_encoder_test.go` line 31 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjQSPVe-8x-4sdADOgK:-NjQSPVe-8x-4sdADOgL:bds3i29) ([raw file](https://github.com/cockroachdb/cockroach/blob/64d770aaae93d666fc480de62cf35c1af2e31843/pkg/server/settingswatcher/setting_encoder_test.go#L31)):*
> ```Go
> )
> 
> func TestSettingsEncoder(t *testing.T) {
> ```

Nice test!


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113996)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @dt and @nvanbenschoten)

___
*[`pkg/server/settingswatcher/setting_encoder.go` line 27 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjQRtu39zT3XZdl7gpa:-NjSiS-E8_RQyxsCu_Rs:b-eeg8gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/64d770aaae93d666fc480de62cf35c1af2e31843/pkg/server/settingswatcher/setting_encoder.go#L27)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we call this from `GetClusterVersionFromStorage`?
</blockquote></details>

I'm tempted to leave these as general incase we run into similar use cases for other settings.
___
*[`pkg/server/settingswatcher/setting_encoder.go` line 29 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjQS1NR2cB8dHEvZqN2:-NjSiP9H0makwhzgyvtI:b-khy73x) ([raw file](https://github.com/cockroachdb/cockroach/blob/64d770aaae93d666fc480de62cf35c1af2e31843/pkg/server/settingswatcher/setting_encoder.go#L29)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Was this intending to pass `setting`, instead of `"version"`?
</blockquote></details>

Yes, I was trying to generalize this code.
___
*[`pkg/sql/set_cluster_setting.go` line 636 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/113996#-NjQSd8F2bsAutu1zc0k:-NjSj5lC6TC0kPET6HBV:blif1di) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf31fbd848ccd58ea3fadd8be33af3fccef6393/pkg/sql/set_cluster_setting.go#L636)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"and extra to"

Something sounds off here.
</blockquote></details>

Ugh..I meant to write:

To achieve the smallest  
//    possible txn and avoid extra operations on other keys, we are going to  
//    use KV call with EncodeSettingKey/EncodeSettingValue functions  
//    instead of using the internal executor.


<!-- Sent from Reviewable.io -->

fqazi commented:
@nvanbenschoten @chrisseto TFTR!

bors r+
